feat(environments): allow all users to start interactive environments

lorenzo-cavazzi opened the pull request:
Allow all users (both unprivileged and anonymous) to start environments on public projects.

* As an unprivileged user, you can start environments also in protected projects and check the pipelines/jobs status.
* As an anonymous user, only the image status will be checked without further details. You are not allowed to ignore that, so it shouldn't be possible to start environments from the wrong image.
A temporary login (not requiring name or any user action) will be triggered when accessing the environments page. This is needed to correctly associate environments to users.

Testing this PR requires an appropriate backend. A preview is available here:
https://andreas.dev.renku.ch

***How to test***
Try to access any public project and start an environment there.

supersede #865 
close #857

rokroskar commented:
what happens if I have a bunch of environments running as an anonymous user but then I follow the advice on the page and log in? I guess I can't find the other environments i started anymore, right?

ableuler commented:
> what happens if I have a bunch of environments running as an anonymous user but then I follow the advice on the page and log in? I guess I can't find the other environments i started anymore, right?

Yes, that ist correct. You'll have them back when you log out though... Would be cool if we could make this transition smoother for users.

ciyer commented on src/api-client/pipeline.js:
```diff
@@ -81,6 +84,44 @@ function addPipelineMethods(client) {
   };
+
+  /**
+   * Get the array of pipeline from GitLab
```
I think this comment is inaccurate (copy/paste?).

ciyer commented on src/api-client/pipeline.js:
```diff
@@ -81,6 +84,44 @@ function addPipelineMethods(client) {
+  };
+
+  /**
+   * Get the array of pipeline from GitLab
```
I think this comment is inaccurate (copy/paste?).

ciyer commented on src/notebooks/Notebooks.present.js:
```diff
@@ -801,6 +846,29 @@ class StartNotebookPipelinesBadge extends Component {
+
+    // unexpected case
+    if (pipelineType !== pipelineTypes.anonymous && pipelineType !== pipelineTypes.logged)
+      return (<div><p>Unexpected state.</p></div>);
```
Can we tell the user what to do in this situation? Reload? Contact an admin?

lorenzo-cavazzi replied:
Good point. Actually, this case should never happen, I added it for extra precaution during the development/test phase but I could remove it completely now (furthermore, there is nothing useful the user can do in that case)

ciyer commented on src/notebooks/Notebooks.state.js:
```diff
@@ -508,6 +522,73 @@ class NotebooksCoordinator {
+          throw error;
+        });
+
+      // TODO: check if filters are oupdated due to discrepancies
```
oupdated -> outdated?

lorenzo-cavazzi replied:
I can remove also this `TODO` since I added somewhere else a line of code to re-trigger the function when something relevant changes (e.g. the commit)

ciyer commented on src/project/Project.present.js:
```diff
@@ -876,46 +834,93 @@ class ProjectEnvironments extends Component {
+        <p>
+          <FontAwesomeIcon icon={faExclamationTriangle} /> As
+          an anonymous user, you can start <ExternalLink role="text" title="Interactive Envirnonments"
+            url="https://renku.readthedocs.io/en/latest/developer/services/notebooks_service.html" /> but
```
` but` -> `, but`

ciyer commented on src/project/Project.present.js:
```diff
@@ -876,46 +834,93 @@ class ProjectEnvironments extends Component {
+      <InfoAlert timeout={0} key="permissions-warning">
+        <p>
+          <FontAwesomeIcon icon={faExclamationTriangle} /> You
+          have limited permissions for this project.
```
Suggestion: 
```You have limited permissions for this project. You can launch an interactive environment, but you will not be able to save any changes. If you want to save your work [...]```

ciyer commented:
I think the information presented here is good, but I think the order should be swapped. Instead of 

### Current
![image](https://user-images.githubusercontent.com/1196411/78676996-7988ac00-78e7-11ea-937d-910b13a2f7fd.png)

### Suggestion

![image](https://user-images.githubusercontent.com/1196411/78677035-8907f500-78e7-11ea-954c-820cf7ea2ecb.png)

And also, instead of "Log in to get all the benefits", how about "Log in for full access."

ciyer commented:
I think the banner here should be the same width as the table:

![image](https://user-images.githubusercontent.com/1196411/78677396-fcaa0200-78e7-11ea-9ec1-fd872cb8d311.png)

ciyer commented:
When the image is not available, I see different information when anonymous or logged in.

When anonymous, it says the build is pending:

![image](https://user-images.githubusercontent.com/1196411/78677956-a38e9e00-78e8-11ea-97af-beedd4a990f6.png)

When logged in, it says there is no image available:

![image](https://user-images.githubusercontent.com/1196411/78678076-c751e400-78e8-11ea-89d1-029accf70256.png)

I realize that it might not be possible to find out if the image is building in anonymous mode, so I think it should always just say the image is not available. We should probably say something like:

"The image for this commit is not currently available. Please [contact a project owner](link to list of project members) for help."

lorenzo-cavazzi commented:
> I think the information presented here is good, but I think the order should be swapped. Instead of
> 
> ### Suggestion
> 
> ![image](https://user-images.githubusercontent.com/1196411/78677035-8907f500-78e7-11ea-954c-820cf7ea2ecb.png)

Do you think the message should be at the end only when there is no running environment or should I always put the message at the end of page? (It will go after the table if there are running environments).

ciyer commented:
> Do you think the message should be at the end only when there is no running environment or should I always put the message at the end of page? (It will go after the table if there are running environments).

Good point. Only if there are no environments running!

lorenzo-cavazzi commented:
Thanks for the thorough review!

I addressed all points, but I ended up with a different (more verbose) explanation for unavailable Docker images. Feel free to suggest something different.

I add the screenshot, I'll update the deployment after I have implemented also the check on `.Values.global.anonymousSessions.enabled`

![Screenshot_20200407_172222](https://user-images.githubusercontent.com/43481553/78688494-9deb8500-78f5-11ea-9a47-7028b6951fef.png)

lorenzo-cavazzi commented:
@ableuler 

I have updated the PR to check for `.Values.global.anonymousSessions.enabled`. When `false`, I assume the Interactive Environments are disabled for anonymous users, therefore I display a message like this

![Screenshot_20200407_194938](https://user-images.githubusercontent.com/43481553/78702730-54a53080-7909-11ea-9809-75e8f0002eb5.png)

Tomorrow I'll try to deploy all the components for supporting unprivileged environments to a new deployment. If you want to update https://andreas.dev.renku.ch, this PR now requires the `tmp-login` endpoint to support the `redirect_url` variable (instead of `redirect-url`)

ableuler commented:
Thanks @lorenzo-cavazzi for the updates to the UI. I have added two deployments for testing (including the latest changes to the UI).
https://tmp-on.dev.renku.ch/ (anonymous sessions enabled)
https://tmp-off.dev.renku.ch/ (anonymous sessions disabled)

lorenzo-cavazzi commented:
Additional feedbacks from @ciyer 
* We may remove the warning message on the Environments list page when at least one environment is running -- no need to keep bothering the user with the same warning again and again
* There are cases when the `/registry/repositories` endpoint return an empty array (e.g. https://tmp-on.dev.renku.ch/projects/andreas/covid-19-dashboard/environments/new ) and the UI keep saying `Checking Docker image status...`
This seems to happen when the repo has no valid Docker mages at all (double check that) -- we can provide the same feedback as when the target image is missing from the registry

rokroskar commented:
I see some repeated requests when accessing the notebook endpoints as a non-authenticated user: 

![image (2)](https://user-images.githubusercontent.com/3353586/78893081-ed06f680-7a6a-11ea-87ab-e297be11e595.png)

lorenzo-cavazzi commented:
> I see some repeated requests when accessing the notebook endpoints as a non-authenticated user [...]

Checking the URL, it seems that there is one repetition. It doesn't seem to come from the UI since it's in `/jupyterhub/.../login-tmp`, while the UI goes to `/api/auth/jupyterhub/.../login-tmp` only once

![Screenshot_20200409_153525](https://user-images.githubusercontent.com/43481553/78903376-53941080-7a7b-11ea-97e8-9c5e01273f35.png)

lorenzo-cavazzi commented:
I fixed the issues found by @ciyer and deployed the new version to https://lorenzotest.dev.renku.ch/

ableuler commented:
> > I see some repeated requests when accessing the notebook endpoints as a non-authenticated user [...]
> 
> Checking the URL, it seems that there is one repetition. It doesn't seem to come from the UI since it's in `/jupyterhub/.../login-tmp`, while the UI goes to `/api/auth/jupyterhub/.../login-tmp` only once
> 
> ![Screenshot_20200409_153525](https://user-images.githubusercontent.com/43481553/78903376-53941080-7a7b-11ea-97e8-9c5e01273f35.png)

Yes, I can confirm what @lorenzo-cavazzi wrote: `/jupyterhub-tmp/hub/api/oauth2/authorize` is hit twice as a part of the regular login flow of the tmp-authenitcator. Also note that `/api/auth/jupyterhub/login-tmp` and `/jupyterhub-tmp/services/notebooks/login-tmp` are two different endpoints.

ableuler commented on src/App.js:
```diff
@@ -52,6 +52,10 @@ class App extends Component {
     }
 
+    // check anonymous sessions settings
+    const anonymousSessions = this.props.params["ANONYMOUS_SESSIONS"] === "true" ? true : false;
```
Is this notation any different from just writing `this.props.params["ANONYMOUS_SESSIONS"] === "true"`?

lorenzo-cavazzi replied:
right, I got used to always specify the 2 values when using the ternary operator to avoid react inline code errors, but it's definitely not necessary here :+1:

ableuler commented on src/App.js:
```diff
@@ -52,6 +52,10 @@ class App extends Component {
 
+    // check anonymous sessions settings
+    const anonymousSessions = this.props.params["ANONYMOUS_SESSIONS"] === "true" ? true : false;
+    const blockAnonymous = !user.logged && !anonymousSessions ? true : false;
```
Same comment as above.

ableuler commented on src/api-client/pipeline.js:
```diff
@@ -63,6 +65,7 @@ function addPipelineMethods(client) {
       method: "GET",
       headers,
     }).then(response => response.data);
+    //return Promise.resolve([{ id: projectId, "status": "success" }]);
```
This was useful for me, guess you can remove it now :)

lorenzo-cavazzi replied:
ops! Good catch!

ableuler commented on src/api-client/pipeline.js:
```diff
@@ -81,6 +84,46 @@ function addPipelineMethods(client) {
+   */
+  client.getRegistries = (projectId) => {
+    const headers = client.getBasicHeaders();
+    headers.append("Content-Type", "application/json");
```
Not necessarily for this PR, but I guess the commonly used headers should be defined as named constants to avoid hunting a typo for half a day at some point.

lorenzo-cavazzi replied:
This makes sense, especially since there are only a couple of headers that are set most of the time. I guess having `ContentType.JSON`, maybe in addition to `client.setContentType()`, would be neat.
I'll create another issue for this.

ableuler replied:
👍

ableuler commented on src/notebooks/Notebooks.present.js:
```diff
@@ -801,6 +868,32 @@ class StartNotebookPipelinesBadge extends Component {
+        <div>
+          <Label>
+            <p>
+              <FontAwesomeIcon icon={faExclamationTriangle} /> The image for this commit is not currently available.
```
```suggestion
              <FontAwesomeIcon icon={faExclamationTriangle} /> The image for this commit is currently not available.
```

ciyer replied:
I prefer the original wording: "The image for this commit is not currently available."

ableuler replied:
👍

ableuler commented on src/notebooks/Notebooks.present.js:
```diff
@@ -801,6 +868,32 @@ class StartNotebookPipelinesBadge extends Component {
+            </p>
+            <p>
+              Since building it takes a while, consider waiting a few minutes if the commit is very recent.
+              <br />Otherwise, you can either select another commit or <ExternalLink role="text" size="sm"
```
```suggestion
              <br />Otherwise, you can either select a previous commit or <ExternalLink role="text" size="sm"
```

lorenzo-cavazzi replied:
I personally think `another` is more precise here. It's possible you tried to select a previous commit that has no valid image, but a more recent one does.

ciyer replied:
I think the original wording is better. It is also possible that the image build on an older commit failed. In that case, the user might want to try out a more recent commit.

ableuler replied:
ok, no problem - fine for me 👍

lorenzo-cavazzi commented on helm-chart/renku-ui/templates/deployment.yaml:
```diff
@@ -44,6 +44,8 @@ spec:
             - name: RENKU_TEMPLATES_REF
               value: {{ required "templateRepository.ref must be specified, e.g., master" .Values.templatesRepository.ref | quote }}
+            - name: ANONYMOUS_SESSIONS
+              value: {{ .Values.global.anonymousSessions.enabled | quote }}
```
I thought about putting a default in the `values` file but, since it's not a UI only setting, I would prefer to put the default value here -- it's not necessary, but I agree it may be easier to read
